config/config.go

```diff
@@ -1545,94 +1547,125 @@ func migrateConfig(configFile, targetDir string) (string, error) {
 	return target, nil
 }
 
-// ReadConfig verifies and checks for encryption and verifies the unencrypted
-// file contains JSON.
-// Prompts for decryption key, if target file is encrypted
+// ReadConfig reads the configuration from the given file
+// if target file is encrypted, prompts for encryption key
+// Also - if not in dryrun mode - it checks if the configuration needs to be encrypted
+// and stores the file as encrypted, if necessary (prompting for enryption key)
 func (c *Config) ReadConfig(configPath string, dryrun bool) error {
 	defaultPath, _, err := GetFilePath(configPath)
 	if err != nil {
 		return err
 	}
-
-	fileData, err := ioutil.ReadFile(defaultPath)
+	confFile, err := os.Open(defaultPath)
 	if err != nil {
 		return err
 	}
+	defer confFile.Close()
+	result, wasEncrypted, err := ReadConfig(confFile, func() ([]byte, error) { return PromptForConfigKey(false) })
+	if err != nil {
+		return fmt.Errorf("error reading config %w", err)
+	}
+	// Override values in the current config
+	*c = *result
 
-	if !ConfirmECS(fileData) {
-		err = json.Unmarshal(fileData, c)
-		if err != nil {
-			return err
-		}
-
-		if c.EncryptConfig == fileEncryptionDisabled {
-			return nil
-		}
+	if dryrun || wasEncrypted || (!wasEncrypted && c.EncryptConfig == fileEncryptionDisabled) {
+		return nil
+	}
 
-		if c.EncryptConfig == fileEncryptionPrompt {
-			confirm, err := promptForConfigEncryption()
-			if err == nil {
-				if confirm {
-					c.EncryptConfig = fileEncryptionEnabled
-					return c.SaveConfig(defaultPath, dryrun)
-				}
+	if c.EncryptConfig == fileEncryptionPrompt {
+		confirm, err := promptForConfigEncryption()
+		if err == nil {
+			if confirm {
+				c.EncryptConfig = fileEncryptionEnabled
+				return c.SaveConfig(defaultPath)
+			}
 
-				c.EncryptConfig = fileEncryptionDisabled
-				err := c.SaveConfig(configPath, dryrun)
-				if err != nil {
-					log.Errorf(log.ConfigMgr, "Cannot save config. Error: %s\n", err)
-				}
+			c.EncryptConfig = fileEncryptionDisabled
+			err := c.SaveConfig(defaultPath)
+			if err != nil {
+				log.Errorf(log.ConfigMgr, "Cannot save config. Error: %s\n", err)
 			}
 		}
-		return nil
 	}
+	return nil
+}
 
-	errCounter := 0
-	for {
-		if errCounter >= maxAuthFailures {
-			return errors.New("failed to decrypt config after 3 attempts")
-		}
-		key, err := PromptForConfigKey(false)
+// ReadConfig verifies and checks for encryption and loads the config.
+// Prompts for decryption key, if target data is encrypted.
+// Returns the loaded configuration and whether it was encrypted.
+func ReadConfig(configReader io.Reader, keyProvider func() ([]byte, error)) (*Config, bool, error) {
```

Comment:
Does this need to be exported?

Reply:
The whole point of this change was to enable loading configuration from non-files and this would be the entry point for these use cases, so this is actually intentionally exported.
Imagine being able to load config from (configuration) database, some custom secure storage, cloud service or a unit test - the ability to not having to create temporary files (and messing with the Stdin for encryption key) just for loading config is useful.

Reply:
Yeah, I don't mind it being exported, it might be nice but out of scope for this; add a quick GRPC endpoint to upload a configuration file and flush all settings to utilise this for future implementations and or maybe even a live reload situation if enabled. This is pretty cool, endless config possibilities.